sdk/core/core-lro/src/poller.ts

```diff
@@ -0,0 +1,133 @@
+import { EventEmitter } from "events";
+import { delay, HttpOperationResponse, RequestOptionsBase, ServiceClient } from "@azure/core-http";
+
+export interface PollerOptionalParameters<States> {
+  manual?: boolean;
+  intervalInMs?: number;
+  state?: States;
+  client?: ServiceClient;
+  initialResponse?: HttpOperationResponse;
+  previousResponse?: HttpOperationResponse;
+  requestOptions?: RequestOptionsBase;
+  stopped?: boolean;
+}
+
+export abstract class Poller<States, Result> extends EventEmitter  {
+  private readonly manual: boolean = false;
+  private stopped: boolean = false;
+  private intervalInMs: number = 1000;
+  private _state: States;
+  protected readonly client: ServiceClient;
+  protected initialResponse?: HttpOperationResponse;
+  protected previousResponse?: HttpOperationResponse;
+  protected requestOptions: RequestOptionsBase | undefined;
+  public stopped: boolean;
+
+  constructor(options: PollerOptionalParameters) {
+    this.manual = Boolean(options.manual);
+    this.intervalInMs = Number(options.intervalInMs);
+    this.client = options.client;
+    this.initialResponse = options.initialResponse;
+    this.previousResponse = options.previousResponse;
+    this.requestOptions = options.requestOptions;
+    this.stopped = Boolean(options.stopped);
+
+    if (options.state) this._state = options.state;
+    if (this.manual) return;
+    this.poll();
+  }
+
+  protected abstract getStateFromResponse(
+    response: HttpOperationResponse
+  ): LongRunningOperationStates;
+
+  protected abstract async sendRequest(options?: RequestOptionsBase): Promise<HttpOperationResponse>;
+
+  protected async processResponse(response: HttpOperationResponse): Promise<void | Result> {
+    if (!this.initialResponse) this.initialResponse = response;
+    this.previousResponse = response;
+    this.emit("newResponse", response, this);
+    this.state = this.getStateFromResponse(response);
```

Comment:
it feels like instead of `getStateFromResponse`, this should call into an abstract method like `handleResponse` which can optionally update state if necessary.

Reply:
We currently have `processResponse` (which can be changed to `handleResponse`). If we let this method be responsible of setting the state, I feel that people will miss it, as well as missing the emission of the event. Let's follow up on teams, I want to make sure I'm understanding your feedback here.

Reply:
`processResponse` is for poller's internal implementation. I want a hook for a subclass of poller to do whatever it needs to do on receiving a message. That might be doing some bookkeeping and updating state, that might be updating the headers for the next request, etc.

I suppose if we want subclasses to be stateless (in order to allow easy rehydration) then it's fine to make the contract about updating state. I'm coming around to this pattern. If setting the state happens externally, then I think we should pass in the previous state, like 

```ts
this.state = this.getStateFromResponse(this.state, response)
```

That way state can be immutable if the subclass wants it to be.

Reply:
Oh, gotcha! I'll do the change.